x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/BufferOnMarkInputStream.java

```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.repositories.encrypted;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+/**
+ * A {@code BufferOnMarkInputStream} adds the mark and reset functionality to another input stream.
+ * All the bytes read or skipped following a {@link #mark(int)} call are also stored in a fixed-size internal array
+ * so they can be replayed following a {@link #reset()} call. The size of the internal buffer is specified at construction
+ * time. It is an error (throws {@code IllegalArgumentException}) to specify a larger {@code readlimit} value as an argument
+ * to a mark call.
+ * <p>
+ * Unlike the {@link java.io.BufferedInputStream} this only buffers upon a {@link #mark(int)} call,
+ * i.e. if {@code mark} is never called this is equivalent to a bare pass-through {@link FilterInputStream}.
+ * Moreover, this does not buffer in advance, so the amount of bytes read from this input stream, at any time, is equal to the amount
+ * read from the underlying stream (provided that reset has not been called, in which case bytes are replayed from the internal buffer
+ * and no bytes are read from the underlying stream).
+ * <p>
+ * Close will also close the underlying stream and any subsequent {@code read}, {@code skip}, {@code available} and
+ * {@code reset} calls will throw {@code IOException}s.
+ * <p>
+ * This is NOT thread-safe, multiple threads sharing a single instance must synchronize access.
+ */
+public final class BufferOnMarkInputStream extends FilterInputStream {
+
+    // all protected for tests
+    protected final int bufferSize;
+    /**
+     * The array used to store the bytes to be replayed upon a reset call.
+     * The buffer portion that stores valid bytes, which must be returned by the read calls after a reset call,
+     * is demarcated by a {@code head} (inclusive) and a {@code tail} offset (exclusive). The offsets wrap around,
+     * i.e. if the {@code tail} offset is smaller than the {@code head} offset, then the portion of valid bytes
+     * is that from the {@code head} offset until the end of the buffer array and from the start of the array
+     * until the {@code tail} offset. The buffer is empty when both the {@code head} and the {@code tail} offsets
+     * are equal. The buffer is full if it stores {@code bufferSize} elements.
+     * To avoid mixing up the two states, the actual allocated size of the array is {@code bufferSize + 1}.
+     */
+    protected byte[] ringBuffer;
```

Comment:
> Firstly, is my understanding correct?

That's correct. This is to support the case of mark after a reset, when the mark happens before reset finishes to replay all the bytes.

This is indeed an optimization but I wouldn't say it's premature, because the ring buffer approach is more performant and does not have adversarial unfavorable cases:

- Two calls to `read(byte[], .... )` instead of a single one is not quite a fair way to weight the trade-off. Two calls to `read(byte[], ... )` eventually translate to two calls to `System.arraycopy` which is exactly the same cost as with the shuffling-of-bytes strategy, i.e. one `System.arraycopy` to move the bytes to the beginning and another one to actually read them. The fixed overhead (i.e. method calls, etc) is different but less important, and also the total lengths of array copies between the two `System.arraycopy` calls is doubly in favor of the ring buffer strategy (the shuffle strategy must move all the bytes that will be read).
- It's not always the case that two `read(byte[], ....)` are required to read the same number of bytes, but it's always the case that a `mark` before the `reset` is done replaying, incurs the shuffling cost. Only a mark-reset-mark-reset case where the byte length between the second mark and reset is large will turn out to require two reads. In other words, only the cases that would require the byte shuffling would possibly require two read calls, but not necessarily.
- the byte shuffling cost is incurred by the `mark` call instead of the `reset` and/or `read` calls and this crosses the expectation from the caller that `mark` is a light call (the expectation is partly due to the fact that `mark` calls are more frequent).

Does all this justify the code complexity? Probably not, given that in practice the performance is pretty much the same and that adversarial cases (mark after every byte read after a reset) are not a threat. But when I embarked on this there was no pressure to get something "good enough" out quickly and I wanted the implementation to be as reusable as possible.

Given that, and that the code is already there, I prefer the third option. I will create an inner class to abstract away the complexity and have documentation explaining _why_ not _how_.

Reply:
The latest commit contains the `BufferOnMarkInputStream` refactored to use a ring buffer inner class.